File: php/uuidgeneratorapi2_sdk.php
<?php
declare(strict_types=1);

// UuidGeneratorApi2 SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class UuidGeneratorApi2SDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new UuidGeneratorApi2Utility();
        $this->_utility = $utility;

        $config = UuidGeneratorApi2Config::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = UuidGeneratorApi2Helpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = UuidGeneratorApi2Helpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, UuidGeneratorApi2Features::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return UuidGeneratorApi2Utility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = UuidGeneratorApi2Helpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = UuidGeneratorApi2Helpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = UuidGeneratorApi2Helpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new UuidGeneratorApi2Spec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = UuidGeneratorApi2Helpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = UuidGeneratorApi2Helpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_guid = null;

    // Idiomatic facade: $client->guid()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Guid() (PHP method
    // names are case-insensitive).
    public function guid($data = null)
    {
        require_once __DIR__ . '/entity/guid_entity.php';
        if ($data === null) {
            if ($this->_guid === null) {
                $this->_guid = new GuidEntity($this, null);
            }
            return $this->_guid;
        }
        return new GuidEntity($this, $data);
    }


    private $_v1n = null;

    // Idiomatic facade: $client->v1n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V1n() (PHP method
    // names are case-insensitive).
    public function v1n($data = null)
    {
        require_once __DIR__ . '/entity/v1n_entity.php';
        if ($data === null) {
            if ($this->_v1n === null) {
                $this->_v1n = new V1nEntity($this, null);
            }
            return $this->_v1n;
        }
        return new V1nEntity($this, $data);
    }


    private $_v1n2 = null;

    // Idiomatic facade: $client->v1n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V1n2() (PHP method
    // names are case-insensitive).
    public function v1n2($data = null)
    {
        require_once __DIR__ . '/entity/v1n2_entity.php';
        if ($data === null) {
            if ($this->_v1n2 === null) {
                $this->_v1n2 = new V1n2Entity($this, null);
            }
            return $this->_v1n2;
        }
        return new V1n2Entity($this, $data);
    }


    private $_v3n = null;

    // Idiomatic facade: $client->v3n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V3n() (PHP method
    // names are case-insensitive).
    public function v3n($data = null)
    {
        require_once __DIR__ . '/entity/v3n_entity.php';
        if ($data === null) {
            if ($this->_v3n === null) {
                $this->_v3n = new V3nEntity($this, null);
            }
            return $this->_v3n;
        }
        return new V3nEntity($this, $data);
    }


    private $_v3n2 = null;

    // Idiomatic facade: $client->v3n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V3n2() (PHP method
    // names are case-insensitive).
    public function v3n2($data = null)
    {
        require_once __DIR__ . '/entity/v3n2_entity.php';
        if ($data === null) {
            if ($this->_v3n2 === null) {
                $this->_v3n2 = new V3n2Entity($this, null);
            }
            return $this->_v3n2;
        }
        return new V3n2Entity($this, $data);
    }


    private $_v4n = null;

    // Idiomatic facade: $client->v4n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V4n() (PHP method
    // names are case-insensitive).
    public function v4n($data = null)
    {
        require_once __DIR__ . '/entity/v4n_entity.php';
        if ($data === null) {
            if ($this->_v4n === null) {
                $this->_v4n = new V4nEntity($this, null);
            }
            return $this->_v4n;
        }
        return new V4nEntity($this, $data);
    }


    private $_v4n2 = null;

    // Idiomatic facade: $client->v4n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V4n2() (PHP method
    // names are case-insensitive).
    public function v4n2($data = null)
    {
        require_once __DIR__ . '/entity/v4n2_entity.php';
        if ($data === null) {
            if ($this->_v4n2 === null) {
                $this->_v4n2 = new V4n2Entity($this, null);
            }
            return $this->_v4n2;
        }
        return new V4n2Entity($this, $data);
    }


    private $_v5n = null;

    // Idiomatic facade: $client->v5n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V5n() (PHP method
    // names are case-insensitive).
    public function v5n($data = null)
    {
        require_once __DIR__ . '/entity/v5n_entity.php';
        if ($data === null) {
            if ($this->_v5n === null) {
                $this->_v5n = new V5nEntity($this, null);
            }
            return $this->_v5n;
        }
        return new V5nEntity($this, $data);
    }


    private $_v5n2 = null;

    // Idiomatic facade: $client->v5n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V5n2() (PHP method
    // names are case-insensitive).
    public function v5n2($data = null)
    {
        require_once __DIR__ . '/entity/v5n2_entity.php';
        if ($data === null) {
            if ($this->_v5n2 === null) {
                $this->_v5n2 = new V5n2Entity($this, null);
            }
            return $this->_v5n2;
        }
        return new V5n2Entity($this, $data);
    }


    private $_v6n = null;

    // Idiomatic facade: $client->v6n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V6n() (PHP method
    // names are case-insensitive).
    public function v6n($data = null)
    {
        require_once __DIR__ . '/entity/v6n_entity.php';
        if ($data === null) {
            if ($this->_v6n === null) {
                $this->_v6n = new V6nEntity($this, null);
            }
            return $this->_v6n;
        }
        return new V6nEntity($this, $data);
    }


    private $_v6n2 = null;

    // Idiomatic facade: $client->v6n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V6n2() (PHP method
    // names are case-insensitive).
    public function v6n2($data = null)
    {
        require_once __DIR__ . '/entity/v6n2_entity.php';
        if ($data === null) {
            if ($this->_v6n2 === null) {
                $this->_v6n2 = new V6n2Entity($this, null);
            }
            return $this->_v6n2;
        }
        return new V6n2Entity($this, $data);
    }


    private $_v7n = null;

    // Idiomatic facade: $client->v7n()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V7n() (PHP method
    // names are case-insensitive).
    public function v7n($data = null)
    {
        require_once __DIR__ . '/entity/v7n_entity.php';
        if ($data === null) {
            if ($this->_v7n === null) {
                $this->_v7n = new V7nEntity($this, null);
            }
            return $this->_v7n;
        }
        return new V7nEntity($this, $data);
    }


    private $_v7n2 = null;

    // Idiomatic facade: $client->v7n2()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias V7n2() (PHP method
    // names are case-insensitive).
    public function v7n2($data = null)
    {
        require_once __DIR__ . '/entity/v7n2_entity.php';
        if ($data === null) {
            if ($this->_v7n2 === null) {
                $this->_v7n2 = new V7n2Entity($this, null);
            }
            return $this->_v7n2;
        }
        return new V7n2Entity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new UuidGeneratorApi2SDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: py/uuidgeneratorapi2_sdk.py
# UuidGeneratorApi2 SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import UuidGeneratorApi2Utility
from core.spec import UuidGeneratorApi2Spec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import UuidGeneratorApi2BaseFeature
from features import _make_feature


class UuidGeneratorApi2SDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = UuidGeneratorApi2Utility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return UuidGeneratorApi2Utility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = UuidGeneratorApi2Spec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Guid(self, data=None) -> "GuidEntity":
        """Entity factory: client.Guid().list() / client.Guid().load({"id": ...})."""
        from entity.guid_entity import GuidEntity
        return GuidEntity(self, data)


    def V1n(self, data=None) -> "V1nEntity":
        """Entity factory: client.V1n().list() / client.V1n().load({"id": ...})."""
        from entity.v1n_entity import V1nEntity
        return V1nEntity(self, data)


    def V1n2(self, data=None) -> "V1n2Entity":
        """Entity factory: client.V1n2().list() / client.V1n2().load({"id": ...})."""
        from entity.v1n2_entity import V1n2Entity
        return V1n2Entity(self, data)


    def V3n(self, data=None) -> "V3nEntity":
        """Entity factory: client.V3n().list() / client.V3n().load({"id": ...})."""
        from entity.v3n_entity import V3nEntity
        return V3nEntity(self, data)


    def V3n2(self, data=None) -> "V3n2Entity":
        """Entity factory: client.V3n2().list() / client.V3n2().load({"id": ...})."""
        from entity.v3n2_entity import V3n2Entity
        return V3n2Entity(self, data)


    def V4n(self, data=None) -> "V4nEntity":
        """Entity factory: client.V4n().list() / client.V4n().load({"id": ...})."""
        from entity.v4n_entity import V4nEntity
        return V4nEntity(self, data)


    def V4n2(self, data=None) -> "V4n2Entity":
        """Entity factory: client.V4n2().list() / client.V4n2().load({"id": ...})."""
        from entity.v4n2_entity import V4n2Entity
        return V4n2Entity(self, data)


    def V5n(self, data=None) -> "V5nEntity":
        """Entity factory: client.V5n().list() / client.V5n().load({"id": ...})."""
        from entity.v5n_entity import V5nEntity
        return V5nEntity(self, data)


    def V5n2(self, data=None) -> "V5n2Entity":
        """Entity factory: client.V5n2().list() / client.V5n2().load({"id": ...})."""
        from entity.v5n2_entity import V5n2Entity
        return V5n2Entity(self, data)


    def V6n(self, data=None) -> "V6nEntity":
        """Entity factory: client.V6n().list() / client.V6n().load({"id": ...})."""
        from entity.v6n_entity import V6nEntity
        return V6nEntity(self, data)


    def V6n2(self, data=None) -> "V6n2Entity":
        """Entity factory: client.V6n2().list() / client.V6n2().load({"id": ...})."""
        from entity.v6n2_entity import V6n2Entity
        return V6n2Entity(self, data)


    def V7n(self, data=None) -> "V7nEntity":
        """Entity factory: client.V7n().list() / client.V7n().load({"id": ...})."""
        from entity.v7n_entity import V7nEntity
        return V7nEntity(self, data)


    def V7n2(self, data=None) -> "V7n2Entity":
        """Entity factory: client.V7n2().list() / client.V7n2().load({"id": ...})."""
        from entity.v7n2_entity import V7n2Entity
        return V7n2Entity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "UuidGeneratorApi2SDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.guid_entity import GuidEntity
    from entity.v1n_entity import V1nEntity
    from entity.v1n2_entity import V1n2Entity
    from entity.v3n_entity import V3nEntity
    from entity.v3n2_entity import V3n2Entity
    from entity.v4n_entity import V4nEntity
    from entity.v4n2_entity import V4n2Entity
    from entity.v5n_entity import V5nEntity
    from entity.v5n2_entity import V5n2Entity
    from entity.v6n_entity import V6nEntity
    from entity.v6n2_entity import V6n2Entity
    from entity.v7n_entity import V7nEntity
    from entity.v7n2_entity import V7n2Entity
